lib/go/thrift/exception.go

```diff
@@ -26,19 +26,86 @@ import (
 // Generic Thrift exception
 type TException interface {
 	error
+
+	TExceptionType() TExceptionType
 }
 
 // Prepends additional information to an error without losing the Thrift exception interface
 func PrependError(prepend string, err error) error {
-	if t, ok := err.(TTransportException); ok {
-		return NewTTransportException(t.TypeId(), prepend+t.Error())
+	msg := prepend + err.Error()
+
+	if te, ok := err.(TException); ok {
+		switch te.TExceptionType() {
+		case TExceptionTypeTransport:
+			if t, ok := err.(TTransportException); ok {
+				return NewTTransportException(t.TypeId(), msg)
+			}
+		case TExceptionTypeProtocol:
+			if t, ok := err.(TProtocolException); ok {
+				return NewTProtocolExceptionWithType(t.TypeId(), errors.New(msg))
+			}
+		case TExceptionTypeApplication:
+			if t, ok := err.(TApplicationException); ok {
+				return NewTApplicationException(t.TypeId(), msg)
+			}
+		}
+
+		return wrappedTException{
+			err:            errors.New(msg),
```

Comment:
@fishy Hi, thank you for expanding thrift errors and Unwrap support!

At this line we are losing here underline error and seems it's OK by design. I'm just curious why we are setting `msg` instead of storing the original err + msg at wrappedTException? 

 I'm working on error handler for thrift clients and the most common case when `context.Canceled` error occurred. The problem is when we perform read\write operations at thrift fields - we can get `context.Canceled` error, but at the moment I can't unwrap thrift result error because of concatenation of pretending msg.

Saving underlying error would be a simplified error handling process. For example:
```golang
// updated wrappedTException
type wrappedTException  struct {
  err            error
  pretendMsg     string
  tExceptionType TExceptionType
}

func (w wrappedTException) Error() string {
  return w.pretendMsg + w.err.Error()
}

// error occursed during thrift ReadField1 and return context.Canceled
func (p *ServiceResult)  ReadField1(iprot thrift.TProtocol) error {
  if err := p.Field.Read(iprot); err != nil {
    return thrift.PrependError(fmt.Sprintf("%T error reading struct: ", p.Field), err)
  }

  return nil
}

// app error helper fn
func isContextCancelledError(err error) bool {
  if err == nil {
    return false
  }

  switch e := err.(type) {
  case thrift.TException:
   return errors.Is(e, context.Canceled)
  default:
    return false
  }
}
```

Reply:
@tte

1. the old `PrependError` doesn't keep the original error either, so this is not a regression :)
2. that being said, this is a reasonable proposal and trivial to implement so I'll just do it
3. I'm curious, which TProtocol/TTransport combination do you use to actually get `context.Canceled` to be wrapped under those `ReadFieldX` functions? I don't think any of the TProtocol implementation would return `context.Canceled` as the error

Reply:
Oh I think this is the one coming from `THeaderTransport`: https://github.com/apache/thrift/blob/7f9abb1cc0f4b2793a48f45ddfcf0d2b287cc50c/lib/go/thrift/header_transport.go#L689, as that's the only place in the codebase we return `ctx.Err()` in any way.

So this is actually `TTransportException` wrapped `context.Canceled`, which means doing it inside `wrappedTException` won't fix the issue. We need to implement that in `tTransportException` as well.

Reply:
but really in the old code (0.13.0) this won't work with prepended `TTransportException` either: https://github.com/apache/thrift/blob/v0.13.0/lib/go/thrift/exception.go#L33-L35

Reply:
@tte See https://github.com/apache/thrift/pull/2308